File: superstruct/planner.py
"""Query planner. The brain of Superstruct.

Three responsibilities sit in this module.

One. Lazy index construction. When a predicate arrives and no built
index can answer it the planner picks the right index type, builds it
from the primary store and registers it.

Two. Cross-structure query decomposition. A query tree of And, Or, Not
and Predicate leaves gets walked. Each leaf is routed to whichever
index can answer it. Inner nodes combine the resulting id sets with
intersection, union or set difference.

Three. Memory budget enforcement. After every build the planner sums
estimated memory across all materialized indexes. If the total exceeds
the budget it evicts the lowest scoring indexes until the total fits.
The workload tracker provides the scores.
"""
import time

from .primary import PrimaryStore, Record
from .query import (
    Query,
    Predicate,
    PredicateKind,
    And,
    Or,
    Not,
    Node,
)
from .indexes.base import Index
from .indexes.hash_index import HashIndex
from .indexes.sorted_index import SortedIndex
from .indexes.trie_index import TrieIndex
from .indexes.inverted_index import InvertedIndex
from .indexes.ngram_index import NgramIndex
from .workload import WorkloadTracker


# Default mapping from predicate kind to the index type best suited to
# answer it. The planner consults this when no built index exists yet
# and it needs to materialize one. EQUALS goes to hash because hash is
# strictly faster than scanning a sorted list. RANGE goes to sorted.
# PREFIX goes to trie. CONTAINS goes to a word level inverted index.
# FUZZY goes to a trigram n gram index.
_DEFAULT_INDEX_FOR_KIND: dict = {
    PredicateKind.EQUALS: HashIndex,
    PredicateKind.RANGE: SortedIndex,
    PredicateKind.PREFIX: TrieIndex,
    PredicateKind.CONTAINS: InvertedIndex,
    PredicateKind.FUZZY: NgramIndex,
}


class Planner:
    """Routes queries to indexes, builds indexes lazily, evicts when full."""

    def __init__(
        self,
        primary: PrimaryStore,
        workload: WorkloadTracker,
        memory_budget_bytes: int,
    ):
        self._primary = primary
        self._workload = workload
        self._memory_budget_bytes = memory_budget_bytes

        # Materialized indexes keyed by (index type name, attribute).
        # This shape means at most one trie per attribute and one
        # sorted index per attribute, which is the right invariant.
        self._indexes: dict[tuple[str, str], Index] = {}

    # -----------------------------------------------------------------
    # Configuration
    # -----------------------------------------------------------------

    def set_memory_budget(self, bytes_limit: int) -> None:
        """Change the memory budget. Triggers an eviction pass right
        away if the current footprint exceeds the new limit."""
        self._memory_budget_bytes = bytes_limit
        self._enforce_memory_budget()

    # -----------------------------------------------------------------
    # Lazy index construction
    # -----------------------------------------------------------------

    def _ensure_index_for(self, predicate: Predicate) -> Index | None:
        """Return an index that can answer this predicate.

        Builds a new index on demand if no existing one can satisfy
        the predicate. Returns None if no index type knows how to
        answer this predicate kind, in which case the caller falls back
        to a primary store scan.
        """
        # Look for an existing materialized index that can answer this
        # predicate. We loop because several index types may be valid
        # and any one of them will do.
        for idx in self._indexes.values():
            if idx.can_answer(predicate):
                return idx

        # No matching index. Pick the default type for this predicate
        # kind. If we have no registered type we return None and let
        # the caller scan the primary store.
        cls = _DEFAULT_INDEX_FOR_KIND.get(predicate.kind)
        if cls is None:
            return None

        # Build and register. We time the build so the workload tracker
        # can give the new index a small loyalty bonus that protects it
        # from immediate eviction right after construction.
        idx = cls(predicate.attribute)
        start = time.monotonic()
        idx.build_from_records(iter(self._primary))
        elapsed = time.monotonic() - start

        key = (cls.__name__, predicate.attribute)
        self._indexes[key] = idx
        self._workload.record_build(key, elapsed)

        # Memory pressure check. If this build pushed us over the budget
        # we drop the cheapest indexes by score until we fit again.
        self._enforce_memory_budget()
        return idx

    def _enforce_memory_budget(self):
        """Drop the lowest scoring indexes until total memory is back
        under the configured budget."""
        total = sum(i.memory_estimate_bytes() for i in self._indexes.values())
        if total <= self._memory_budget_bytes:
            return

        # Sort keys ascending by score. Lowest score evicts first. We
        # snapshot the keys list because we mutate the dict in the loop.
        keys_sorted = sorted(
            list(self._indexes.keys()),
            key=lambda k: self._workload.score(k),
        )
        for key in keys_sorted:
            if total <= self._memory_budget_bytes:
                return
            idx = self._indexes.pop(key)
            total -= idx.memory_estimate_bytes()
            # Forget the stats too. A future rebuild starts from cold
            # and earns its place again.
            self._workload.forget(key)

    # -----------------------------------------------------------------
    # Query execution. Cross-structure decomposition lives here.
    # -----------------------------------------------------------------

    def execute(self, query: Query) -> list[Record]:
        """Run a Query end to end. Returns matching records."""
        # Empty query means every record. Skip predicate routing.
        if query.where is None and query.top_k is None:
            return list(self._primary)

        # Resolve the candidate id set.
        if query.where is None:
            # No predicates but a top_k step exists. Universe is every
            # record in the primary store.
            candidate_ids = {r.id for r in self._primary}
        else:
            candidate_ids = self._evaluate_node(query.where)

        if query.top_k is not None:
            # Final ordering. We hydrate to records here so we can sort
            # by the requested attribute. Skip records that lack the
            # ordering attribute since None and numbers cannot compare.
            attr = query.top_k.attribute
            records = [self._primary.get(rid) for rid in candidate_ids]
            records = [
                r for r in records
                if r is not None and attr in r.attrs
            ]
            records.sort(
                key=lambda r: r.attrs[attr],
                reverse=query.top_k.descending,
            )
            return records[: query.top_k.k]

        # No top-k. Hydrate ids to records in stable id order so test
        # output is deterministic.
        records = [self._primary.get(rid) for rid in sorted(candidate_ids)]
        return [r for r in records if r is not None]

    def _evaluate_node(self, node: Node) -> set[int]:
        """Recursively walk the AST and return the set of matching ids.

        Predicate is the leaf case and dispatches to an index. And, Or
        and Not combine child results with intersection, union and
        complement against the universe of all primary record ids.
        """
        if isinstance(node, Predicate):
            return self._evaluate_predicate(node)
        if isinstance(node, And):
            return self._evaluate_and(node)
        if isinstance(node, Or):
            return self._evaluate_or(node)
        if isinstance(node, Not):
            return self._evaluate_not(node)

        # Defensive default. The Node type alias should keep this
        # branch unreachable in well formed queries.
        return set()

    def _evaluate_predicate(self, pred: Predicate) -> set[int]:
        """Route a single predicate to its index and return ids."""
        idx = self._ensure_index_for(pred)
        if idx is None:
            # No index type registered for this predicate kind. Fall
            # back to a primary store scan. Slow but always correct.
            return self._scan(pred)
        ids = idx.execute(pred)
        self._workload.record_hit((type(idx).__name__, idx.attribute))
        return ids

    def _evaluate_and(self, node: And) -> set[int]:
        """Intersect child id sets. Empty child list means every record."""
        if not node.children:
            return {r.id for r in self._primary}

        result: set[int] | None = None
        for child in node.children:
            child_ids = self._evaluate_node(child)
            if result is None:
                result = child_ids
            else:
                result &= child_ids
            # Empty intersection short circuits. The conjunctive AND
            # cannot grow back from empty.
            if not result:
                return set()
        return result or set()

    def _evaluate_or(self, node: Or) -> set[int]:
        """Union child id sets. Empty child list means no records."""
        out: set[int] = set()
        for child in node.children:
            out |= self._evaluate_node(child)
        return out

    def _evaluate_not(self, node: Not) -> set[int]:
        """Complement of the child against the universe of all records."""
        if node.child is None:
            # Empty NOT means nothing is excluded so every record matches.
            return {r.id for r in self._primary}
        universe = {r.id for r in self._primary}
        excluded = self._evaluate_node(node.child)
        return universe - excluded

    def _scan(self, predicate: Predicate) -> set[int]:
        """Fallback for predicates whose kind has no registered index."""
        out: set[int] = set()
        for r in self._primary:
            v = r.attrs.get(predicate.attribute)
            if v is None:
                continue
            if predicate.kind is PredicateKind.EQUALS and v == predicate.value:
                out.add(r.id)
        return out

    # -----------------------------------------------------------------
    # Maintenance
    # -----------------------------------------------------------------

    def on_insert(self, record: Record) -> None:
        """Tell every materialized index about a new record.

        Indexes that have not been built yet do not need notification.
        They will pick the record up the next time they are built.
        """
        for idx in self._indexes.values():
            idx.insert(record)

    def on_delete(self, record: Record) -> None:
        """Tell every materialized index about a removed record."""
        for idx in self._indexes.values():
            idx.remove(record)

    def index_inventory(self) -> list[tuple[str, str, int]]:
        """Snapshot of currently materialized indexes.

        One tuple per index containing the type name, the attribute
        and an estimated byte size. Useful for the demo and for any
        introspective UI.
        """
        return [
            (type(i).__name__, i.attribute, i.memory_estimate_bytes())
            for i in self._indexes.values()
        ]

File: src/planner.rs
use crate::index::base::Index;
use crate::index::{HashIndex, SortedIndex, TrieIndex, InvertedIndex, NgramIndex};
use crate::primary::{PrimaryStore, Record};
use crate::query::{Predicate, PredicateKind, Node, And, Or, Not, Query};
use crate::value::Attrs;
use crate::workload::WorkloadTracker;
use std::collections::{HashMap, HashSet};
use std::time::Instant;

type IndexKey = (String, String);

fn build_index_for(kind: PredicateKind, attribute: &str) -> Box<dyn Index> {
    match kind {
        PredicateKind::Equals => Box::new(HashIndex::new(attribute.to_string())),
        PredicateKind::Range => Box::new(SortedIndex::new(attribute.to_string())),
        PredicateKind::Prefix => Box::new(TrieIndex::new(attribute.to_string())),
        PredicateKind::Contains => Box::new(InvertedIndex::new(attribute.to_string())),
        PredicateKind::Fuzzy => Box::new(NgramIndex::new(attribute.to_string())),
    }
}

fn index_type_name(kind: PredicateKind) -> &'static str {
    match kind {
        PredicateKind::Equals => "HashIndex",
        PredicateKind::Range => "SortedIndex",
        PredicateKind::Prefix => "TrieIndex",
        PredicateKind::Contains => "InvertedIndex",
        PredicateKind::Fuzzy => "NgramIndex",
    }
}

pub struct Planner {
    indexes: HashMap<IndexKey, Box<dyn Index>>,
    memory_budget_bytes: usize,
}

impl Planner {
    pub fn new(memory_budget_bytes: usize) -> Self {
        Self {
            indexes: HashMap::new(),
            memory_budget_bytes,
        }
    }

    pub fn set_memory_budget(&mut self, bytes_limit: usize, workload: &mut WorkloadTracker) {
        self.memory_budget_bytes = bytes_limit;
        self.enforce_memory_budget(workload);
    }

    fn enforce_memory_budget(&mut self, workload: &mut WorkloadTracker) {
        let mut total: usize = self.indexes.values().map(|i| i.memory_estimate_bytes()).sum();
        if total <= self.memory_budget_bytes {
            return;
        }

        let mut keys: Vec<IndexKey> = self.indexes.keys().cloned().collect();
        keys.sort_by(|a, b| {
            workload
                .score(a)
                .partial_cmp(&workload.score(b))
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        for key in &keys {
            if total <= self.memory_budget_bytes {
                break;
            }
            if let Some(idx) = self.indexes.remove(key) {
                total = total.saturating_sub(idx.memory_estimate_bytes());
                workload.forget(key);
            }
        }
    }

    fn ensure_index_for(
        &mut self,
        predicate: &Predicate,
        primary: &PrimaryStore,
        workload: &mut WorkloadTracker,
    ) -> IndexKey {
        for idx in self.indexes.values() {
            if idx.can_answer(predicate) {
                let kind_name = index_type_name(predicate.kind);
                return (kind_name.to_string(), predicate.attribute.clone());
            }
        }

        let attr = predicate.attribute.clone();
        let kind_name = index_type_name(predicate.kind);
        let mut idx = build_index_for(predicate.kind, &attr);

        let records: Vec<Record> = primary.iter().cloned().collect();
        let start = Instant::now();
        idx.build_from_records(&records);
        let elapsed = start.elapsed().as_secs_f64();

        let key: IndexKey = (kind_name.to_string(), attr);
        workload.record_build(key.clone(), elapsed);
        self.indexes.insert(key.clone(), idx);
        self.enforce_memory_budget(workload);
        key
    }

    fn get_index(&self, key: &IndexKey) -> Option<&dyn Index> {
        self.indexes.get(key).map(|b| b.as_ref())
    }

    pub fn execute(
        &mut self,
        query: &Query,
        primary: &PrimaryStore,
        workload: &mut WorkloadTracker,
    ) -> Vec<Attrs> {
        match (&query.r#where, &query.top_k) {
            (None, None) => primary.iter().map(|r| r.attrs.clone()).collect(),
            (None, Some(tk)) => {
                let ids: HashSet<u64> = primary.ids().collect();
                self.apply_top_k(ids, tk, primary)
            }
            (Some(where_clause), _) => {
                let ids = self.evaluate_node(where_clause, primary, workload);
                if let Some(tk) = &query.top_k {
                    self.apply_top_k(ids, tk, primary)
                } else {
                    let mut sorted: Vec<u64> = ids.into_iter().collect();
                    sorted.sort_unstable();
                    sorted
                        .iter()
                        .filter_map(|id| primary.get(*id))
                        .map(|r| r.attrs.clone())
                        .collect()
                }
            }
        }
    }

    fn apply_top_k(
        &self,
        ids: HashSet<u64>,
        top_k: &crate::query::TopK,
        primary: &PrimaryStore,
    ) -> Vec<Attrs> {
        let attr = &top_k.attribute;
        let mut records: Vec<(&Attrs, &crate::Value)> = ids
            .iter()
            .filter_map(|id| primary.get(*id))
            .filter_map(|r| r.attrs.get(attr).map(|v| (&r.attrs, v)))
            .collect();

        records.sort_by(|(_, a), (_, b)| {
            let cmp = a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal);
            if top_k.descending {
                cmp.reverse()
            } else {
                cmp
            }
        });

        records
            .into_iter()
            .take(top_k.k)
            .map(|(attrs, _)| attrs.clone())
            .collect()
    }

    fn evaluate_node(
        &mut self,
        node: &Node,
        primary: &PrimaryStore,
        workload: &mut WorkloadTracker,
    ) -> HashSet<u64> {
        match node {
            Node::Predicate(p) => self.evaluate_predicate(p, primary, workload),
            Node::And(a) => self.evaluate_and(a, primary, workload),
            Node::Or(o) => self.evaluate_or(o, primary, workload),
            Node::Not(n) => self.evaluate_not(n, primary, workload),
        }
    }

    fn evaluate_predicate(
        &mut self,
        pred: &Predicate,
        primary: &PrimaryStore,
        workload: &mut WorkloadTracker,
    ) -> HashSet<u64> {
        let key = self.ensure_index_for(pred, primary, workload);
        match self.get_index(&key) {
            Some(idx) => {
                workload.record_hit(key);
                idx.execute(pred)
            }
            None => self.scan(pred, primary),
        }
    }

    fn evaluate_and(
        &mut self,
        node: &And,
        primary: &PrimaryStore,
        workload: &mut WorkloadTracker,
    ) -> HashSet<u64> {
        if node.children.is_empty() {
            return primary.ids().collect();
        }
        let mut children = node.children.iter();
        let mut result = self.evaluate_node(children.next().unwrap(), primary, workload);
        for child in children {
            let child_ids = self.evaluate_node(child, primary, workload);
            result = result.intersection(&child_ids).copied().collect();
            if result.is_empty() {
                break;
            }
        }
        result
    }

    fn evaluate_or(
        &mut self,
        node: &Or,
        primary: &PrimaryStore,
        workload: &mut WorkloadTracker,
    ) -> HashSet<u64> {
        let mut out = HashSet::new();
        for child in &node.children {
            out.extend(self.evaluate_node(child, primary, workload));
        }
        out
    }

    fn evaluate_not(
        &mut self,
        node: &Not,
        primary: &PrimaryStore,
        workload: &mut WorkloadTracker,
    ) -> HashSet<u64> {
        let universe: HashSet<u64> = primary.ids().collect();
        match &node.child {
            None => universe,
            Some(child) => {
                let excluded = self.evaluate_node(child, primary, workload);
                universe.difference(&excluded).copied().collect()
            }
        }
    }

    fn scan(&self, predicate: &Predicate, primary: &PrimaryStore) -> HashSet<u64> {
        let mut out = HashSet::new();
        for record in primary.iter() {
            if let Some(v) = record.attrs.get(&predicate.attribute) {
                if predicate.kind == PredicateKind::Equals && v == &predicate.value {
                    out.insert(record.id);
                }
            }
        }
        out
    }

    pub fn on_insert(&mut self, record: &Record) {
        for idx in self.indexes.values_mut() {
            idx.insert(record);
        }
    }

    pub fn on_delete(&mut self, record: &Record) {
        for idx in self.indexes.values_mut() {
            idx.remove(record);
        }
    }

    pub fn index_inventory(&self) -> Vec<(String, String, usize)> {
        self.indexes
            .iter()
            .map(|(k, idx)| {
                (k.0.clone(), idx.attribute().to_string(), idx.memory_estimate_bytes())
            })
            .collect()
    }
}
